gdal/ogr/ogrsf_frmts/dwg/ogrdgnv8layer.cpp

```diff
@@ -853,7 +886,7 @@ std::vector<tPairFeatureHoleFlag> OGRDGNV8Layer::ProcessElement(
                     break;
                     case 0x1995: // 0x1995 (6549) IPCC/Portugal
                     {
-                        char *pszAsSwappedHex = CPLBigEndianBinaryToHex( (OdUInt32)pabyData.size(), (GByte*) pabyData.asArrayPtr() );
+                        char *pszAsSwappedHex = BigEndianBinaryToHex( (OdUInt32)pabyData.size(), (GByte*) pabyData.asArrayPtr() );
```

Comment:
sorry to insist, but I still don't like the current output. For me it should be either an array of decoded integers [ 1234, 4567, ... ] (I'd lean towards that if the bytes are effictively always grouped by 16 bits) or a string with explicit 16-bit hexadecimal number "0xABCD 0xEF12" . "ABCDEF12" with this "big endian" order is higly unusual.

Reply:
@rouault You don't give up, neither I do. That's good! I'll try to gather all my arguments, for a good discussion :-)

As explained in the description, this PR was submitted to solve a use case. The goal was to use GDAL in a workflow to transform old CAD based cartography into GIS based cartography. GDAL is the perfect tool for the job.

### Context 
Until recently, Portuguese official cartography was produced in CAD based systems, using a multi-codification system. Each element in CAD (features) have their type defined by a list of codes. As required by Bentley, these codes were recorded as user data linkages, under a well known application id. The application id used by IPCC (the former map agency) is 0x1995. Companies and developers were able to build applications knowing the information structure encoded in this specific user data linkages. The structure comprises the domain, subdomain, family and object.

The structure is (values in hex):
<table>
  <tr>
    <td>1st word</td>
    <td>1</td>
    <td>0</td>
    <td>0</td>
    <td>3</td>
  </tr>
 <tr>
    <td>2nd word</td>
    <td>1</td>
    <td>9</td>
    <td>9</td>
    <td>5</td>
  </tr>
<tr>
    <td>3rd word</td>
    <td colspan="2">domain</td>
    <td colspan="2">subdomain</td>
</tr>
<tr>
    <td>4th word</td>
    <td colspan="2">family</td>
    <td colspan="2">object</td>
</tr>
</table>
The Microstation print screen provided in the PR shows this user data linkage as displayed to the user, in the CAD environment.

### Purpose of the PR 
This PR was submitted to add the ability to read data from user data linkages from known application ids and report it and a new feature attribute. The idea was to identify the target application id as an environment variable and these drivers would use it to extract the information for that specific user data linkage.

This PR improves two drivers: the DGN and the DGNV8 driver. For the DGNV8 driver, we can take advantage of ODA API to retrieve data for known application ids. For DGN driver, we have to add the necessary logic to retrieve the information from the data, according to each application id.

For unknown applications ids, we are not able to retrieve the data. We just know the size, and from the DGN user data linkage format, that size is always a multiple of 4 words, but its contents can be anything.

This additional capability of these drivers does not affect any previous functionality. Only adds an additional capability to read known user data linkages.

### PR improvement
Based on your suggestion, instead of just reporting one user data linkage type, these two drivers could report all user data linkage present. 

For unknown application ids, that is not possible. The solution discussed here, which I agreed, was to output it as raw data, either as hex or integers. Any dump format for me is ok, because we really don't know what we are dealing with. We just know the size of the data, and that the size is multiple of 4 words. Is up to the driver's user to use this raw data to recover the information encoded on it.

### First PR limitation
Based on your comments, I gave up on DGN driver. But I was so committed to moving forward, that I've accepted to make it happen just for the DGNV8 driver. Right now, I regret that decision, because all the users I know that might want to take advantage of this do not have access to the DGNV8 driver.

### Reverting the PR goal
I'm sorry that that you don't like the output of the application id 0x1995, even for the DGNV8 driver.

For unknown application ids, I agree that we can use any kind of raw data that you like, but for application ids that we know how the structure is, it does not make any sense to me to report it in a different way. The "highly unusual order", as you call it, is the exact order that is displayed in the Microstation [print screen](https://user-images.githubusercontent.com/163681/96733482-6932ad00-13b1-11eb-82d4-c47891bae1f2.png), if you not have noticed. Users don't know any order other than this "highly unusual order".  

Reporting its contents without attending its format, reverts the purpose of this PR. The purpose was to recover the data on that user data linkage. 

### Moving forward
Right now, after a few commits, I've lost completely the purpose of  the PR. Is is not useful any more for the use case and for the users that would like to process old CAD based cartography.

#### Back to the origin
My humble suggestion is to go back to the very first commit and discuss the ability to support the user data linkage of the application id = 0x1995, if requested by the user, by providing a environment variable. Just that specific user data linkage, but attending its structure. 

This support does not interfere any other drivers capability; does not add additional noise if the environment variable is not set, etc. Seems quite inoffensive for me, in terms of impact to the code. And since we don't have many option to deal wit hit, we don't have many things to discuss.

It does not solve all the unforeseen useful use cases, but solves one use case.

Can we restart this from the beginning? What do you think?

(Sorry for such a long comment)

Reply:
> Can we restart this from the beginning? What do you think?

On the contrary, I think we're quite close to what I'd wish. I like the output of Microstation with the 0x prefix. That's exactly what I suggested above: "0x3412 0x7856" as a string has no ambiguity whereas "34127856" is confusing, if the raw bytes where actually ordered as 0x12,0x34,0x56,0x78. Said otherwise if you present the user data link content as 16 bit words, this structure must clearly appear.
I'm only speaking about presentation of raw bytes. The more structured output for DGNv8 for known user data links is fine.

Reply:
Thank you @rouault for the feedback.

## ULink format

### DGN ULink format

The Ulink is a JSON string, mapping application ids and its contents. The content is an array of words (16bits) in hexadecimal format. 
```json
{ "65": [ { "raw": [ "0x1007", "0x0041", "0x0802", "0x0001", "0x0053", "0x0000", "0x0000", "0x0000" ], "size": 16, "type": "unknown" } ] }
```
If the application id is known, the attribute type reports its.
```json
{ "24721": [ { "raw": [ "0x1007", "0x6091", "0x0f81", "0x0021", "0x0964", "0x0000", "0x0000", "0x0000" ], "size": 16, "type": "Oracle" } ] }
```

### DGNV8 ULink format

The ULink is a JSON string, mapping application ids and its contents. The content of the user data linkage is reported for some specific application ids.
```json
{ "24721": [ { "size": 12, "tableId": 33, "MSLink": 2404, "type": "Oracle" } ] }
```
For unknown application ids, content is reported as raw 16bit words (exactly as in the DGN driver).
```json
{ "65": [ { "size": 12, "raw": [ "0x0008", "0x0000", "0x0053", "0x0000", "0x0000", "0x0000" ], "type": "unknown" } ] }
```

## Other examples

### Test file `franck_without_enhanced_precision.dgn`

Example using CAD file used on [ticket 1387](https://trac.osgeo.org/gdal/ticket/1387) 

### DGN driver output

```
OGRFeature(elements):39
  Type (Integer) = 4
  Level (Integer) = 3
  GraphicGroup (Integer) = 0
  ColorIndex (Integer) = 3
  Weight (Integer) = 0
  Style (Integer) = 0
  EntityNum (Integer) = 33
  MSLink (Integer) = 2404
  ULink (String(JSON)) = { "24721": [ { "raw": [ "0x1007", "0x6091", "0x0f81", "0x0021", "0x0964", "0x0000", "0x0000", "0x0000" ], "size": 16, "type": "Oracle" } ] }
  Style = PEN(id:"ogr-pen-0",c:#ff0000)
  LINESTRING (659669.489 216455.466,659642.447 216461.671,659624.602 216463.187)
```
The ULink is a map between application ids and its contents. Since we dump the contents of the user data linkage data, **the second word of the `"raw"` array (in hex) is equal to the object's key (in decimal)**. In the above case, the object key `"27721" =  "0x6091"`.

### DGNV8 driver output

The same previous element, dumped with DGNV8 is:
```
OGRFeature(Default):193
  Type (Integer) = 4
  Level (Integer) = 3
  GraphicGroup (Integer) = 0
  ColorIndex (Integer) = 3
  Weight (Integer) = 0
  Style (Integer) = 0
  ULink (String(JSON)) = { "24721": [ { "size": 12, "tableId": 33, "MSLink": 2404, "type": "Oracle" } ] }
  Style = PEN(id:"ogr-pen-0",c:#ff0000)
  LINESTRING (659669.489 216455.466,659642.447 216461.671,659624.602 216463.187)
```

### Test file `454_1_10k_MNT.dgn`

### DGN driver output (just the ULink attribute)

```json
ULink (String(JSON)) = { "6549": [ { "raw": [ "0x1003", "0x1995", "0x0203", "0x0104" ], "size": 8, "type": "IPCC\/Portugal" }, { "raw": [ "0x1003", "0x1995", "0x1001", "0x0404" ], "size": 8, "type": "IPCC\/Portugal" } ], "32047": [ { "raw": [ "0x1003", "0x7d2f", "0x712e", "0x0005" ], "size": 8, "type": "ODBC" } ] }
```

### DGNV8 driver output (just the ULink attribute)

```json
ULink (String(JSON)) = { "6549": [ { "size": 4, "domain": "0x02", "subdomain": "0x01", "family": "0x01", "object": "0x05", "key": "02010105", "type": "IPCC\/Portugal" }, { "size": 4, "domain": "0x02", "subdomain": "0x01", "family": "0x01", "object": "0x06", "key": "02010106", "type": "IPCC\/Portugal" }, { "size": 4, "domain": "0x02", "subdomain": "0x01", "family": "0x01", "object": "0x07", "key": "02010107", "type": "IPCC\/Portugal" }, { "size": 4, "domain": "0x11", "subdomain": "0x10", "family": "0x01", "object": "0x02", "key": "11100102", "type": "IPCC\/Portugal" }, { "size": 4, "domain": "0x88", "subdomain": "0x01", "family": "0x01", "object": "0x04", "key": "88010104", "type": "IPCC\/Portugal" } ], "24162": [ { "size": 12, "tableId": 5, "MSLink": 2010106, "type": "ODBC" }, { "size": 12, "tableId": 5, "MSLink": 2010107, "type": "ODBC" } ] }
```